File: Data Structures & Algorithms/reorder-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:

        fastPointer = head
        slowPointer = head

        while fastPointer.next and fastPointer.next.next:
            fastPointer = fastPointer.next.next
            slowPointer = slowPointer.next
        
        prev=None
        secondhalf = slowPointer.next
        slowPointer.next = None
        curr = secondhalf

        while curr:
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
        
        first = head
        second = prev

        while second:
            tmp1 = first.next
            tmp2 = second.next

            first.next = second
            second.next = tmp1

            first = tmp1
            second = tmp2

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-4.cpp
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode* dummy = new ListNode(0, head);
        ListNode* lastSection = dummy;

        while (true) {

            ListNode* curr = lastSection->next;
            int count = 0;
            while (curr != nullptr && count < k) {
                curr = curr->next;
                count++;
            }

            if (count < k) {
                break;
            }

            ListNode* curr1 = lastSection->next;
            ListNode* prev = curr; 
            
            while (curr1 != curr) {
                ListNode* saved = curr1->next;
                curr1->next = prev;
                prev = curr1;
                curr1 = saved;
            }

            ListNode* oldGroupStart = lastSection->next;

            lastSection->next = prev;

            lastSection = oldGroupStart;
        }

        return dummy->next;
    }
};